libvips/foreign/spngload.c

```diff
@@ -0,0 +1,800 @@
+
+#ifdef DEBUG
+	printf( "vips_foreign_load_png_generate: line %d, %d rows\n", 
+		r->top, r->height );
+	printf( "vips_foreign_load_png_generate: y_top = %d\n", png->y_pos );
+#endif /*DEBUG*/
+
+	/* We're inside a tilecache where tiles are the full image width, so
+	 * this should always be true.
+	 */
+	g_assert( r->left == 0 );
+	g_assert( r->width == or->im->Xsize );
+	g_assert( VIPS_RECT_BOTTOM( r ) <= or->im->Ysize );
+
+	/* Tiles should always be a strip in height, unless it's the final
+	 * strip.
+	 */
+	g_assert( r->height == VIPS_MIN( VIPS__FATSTRIP_HEIGHT, 
+		or->im->Ysize - r->top ) ); 
+
+	/* And check that y_pos is correct. It should be, since we are inside
+	 * a vips_sequential().
+	 */
+	if( r->top != png->y_pos ) {
+		vips_error( class->nickname, 
+			_( "out of order read at line %d" ), png->y_pos );
+		return( -1 );
+	}
+
+	for( y = 0; y < r->height; y++ ) {
+		error = spng_decode_row( png->ctx, 
+			VIPS_REGION_ADDR( or, 0, r->top + y ),
+			VIPS_REGION_SIZEOF_LINE( or ) );
+		/* libspng returns EOI when successfully reading the 
+		 * final line of input.
+		 */
+		if( error != 0 &&
+			error != SPNG_EOI ) { 
+			/* We've failed to read some pixels. Knock this 
+			 * operation out of cache. 
+			 */
+			vips_operation_invalidate( VIPS_OPERATION( png ) ); 
+
+#ifdef DEBUG
+			printf( "vips_foreign_load_png_generate:\n" ); 
+			printf( "  spng_decode_row() failed, line %d\n", 
+				r->top + y ); 
+			printf( "  thread %p\n", g_thread_self() );
+			printf( "  error %s\n", spng_strerror( error ) ); 
+#endif /*DEBUG*/
+
+			/* And bail if fail is on. 
+			 */
+			if( load->fail ) {
+				vips_error( class->nickname, 
+					"%s", _( "libpng read error" ) ); 
+				return( -1 );
+			}
+		}
+
+		png->y_pos += 1;
+	}
+
+	return( 0 );
+}
+
+static int
+vips_foreign_load_png_load( VipsForeignLoad *load )
+{
+	VipsObjectClass *class = VIPS_OBJECT_GET_CLASS( load );
+	VipsForeignLoadPng *png = (VipsForeignLoadPng *) load;
+	VipsImage **t = (VipsImage **) 
+		vips_object_local_array( VIPS_OBJECT( load ), 3 );
+
+	int error;
+
+	if( vips_source_decode( png->source ) )
+		return( -1 );
+
+	if( png->ihdr.interlace_method != SPNG_INTERLACE_NONE ) {
+		/* Arg awful interlaced image. We have to load to a huge mem 
+		 * buffer, then copy to out.
+		 */
+		t[0] = vips_image_new_memory();
+		vips_foreign_load_png_set_header( png, t[0] );
+		if( vips_image_write_prepare( t[0] ) )
+			return( -1 );
+
+		if( (error = spng_decode_image( png->ctx, 
+			VIPS_IMAGE_ADDR( t[0], 0, 0 ), 
+			VIPS_IMAGE_SIZEOF_IMAGE( t[0] ), 
+			png->fmt, 0 )) ) {
+			vips_error( class->nickname, 
+				"%s", spng_strerror( error ) ); 
+			return( -1 );
+		}
+
+		if( vips_image_write( t[0], load->real ) )
+			return( -1 );
+	}
+	else {
+		t[0] = vips_image_new();
+		vips_foreign_load_png_set_header( png, t[0] );
+
+		/* Initialize for progressive decoding.
+		 */
+		if( (error = spng_decode_image( png->ctx, NULL, 0, 
+			png->fmt, SPNG_DECODE_PROGRESSIVE )) ) {
```

Comment:
Should be `SPNG_DECODE_PROGRESSIVE | SPNG_DECODE_TRNS`, the pixels specified to be transparent in the tRNS chunk will not be transparent without this flag, should also do this for the other `spng_decode_image()` call.

Reply:
fixed with f96f2d30